File: frontend/src/utils/permissions.js
/**
 * Permission Utilities
 *
 * Frontend uchun role-based permission checking.
 * Backend'dagi User model permission metodlari bilan mos.
 */

// Rollar (backend bilan bir xil)
export const ROLES = {
  EMPLOYEE: 'employee',
  MANAGER: 'manager',
  ACCOUNTANT: 'accountant',
  CHIEF_ACCOUNTANT: 'chief_accountant',
  SUPERUSER: 'superuser',

  // Legacy support
  ADMIN: 'admin',
  HR: 'hr',
  VIEWER: 'viewer',
}

// Role display nomlar
export const ROLE_LABELS = {
  [ROLES.EMPLOYEE]: 'Oddiy xodim',
  [ROLES.MANAGER]: "Bo'lim boshlig'i",
  [ROLES.ACCOUNTANT]: 'Buxgalter',
  [ROLES.CHIEF_ACCOUNTANT]: 'Glavniy buxgalter',
  [ROLES.SUPERUSER]: 'Administrator',
  [ROLES.ADMIN]: 'Administrator',
  [ROLES.HR]: 'HR Manager',
  [ROLES.VIEWER]: "Ko'ruvchi",
}

// Role ranglar (badge uchun)
export const ROLE_COLORS = {
  [ROLES.EMPLOYEE]: 'gray',
  [ROLES.MANAGER]: 'blue',
  [ROLES.ACCOUNTANT]: 'green',
  [ROLES.CHIEF_ACCOUNTANT]: 'purple',
  [ROLES.SUPERUSER]: 'red',
  [ROLES.ADMIN]: 'red',
  [ROLES.HR]: 'indigo',
  [ROLES.VIEWER]: 'gray',
}

/**
 * Foydalanuvchining roli borligini tekshirish
 *
 * @param {Object} user - Foydalanuvchi obyekti
 * @param {string|string[]} roles - Tekshiriladigan rol(lar)
 * @returns {boolean}
 */
export const hasRole = (user, roles) => {
  if (!user || !user.role) return false

  const roleArray = Array.isArray(roles) ? roles : [roles]
  return roleArray.includes(user.role)
}

/**
 * Foydalanuvchi permission ga ega yoki yo'qligini tekshirish
 *
 * Backend'dagi has_permission() metodi bilan mos.
 *
 * @param {Object} user - Foydalanuvchi obyekti
 * @param {string} permission - Permission nomi
 * @returns {boolean}
 */
export const hasPermission = (user, permission) => {
  if (!user || !user.role) return false

  // Superuser har doim ruxsat oladi
  if (user.role === ROLES.SUPERUSER || user.role === ROLES.ADMIN) {
    return true
  }

  // Permission mapping (backend bilan bir xil)
  const permissionMap = {
    // Legacy permissions
    can_manage_users: [ROLES.ADMIN, ROLES.SUPERUSER],
    can_edit_employees: [ROLES.ADMIN, ROLES.HR, ROLES.SUPERUSER],
    can_view_reports: [
      ROLES.ADMIN, ROLES.HR, ROLES.VIEWER,
      ROLES.MANAGER, ROLES.ACCOUNTANT,
      ROLES.CHIEF_ACCOUNTANT, ROLES.SUPERUSER
    ],
    can_manage_devices: [ROLES.ADMIN, ROLES.SUPERUSER],
    can_sync_attendance: [ROLES.ADMIN, ROLES.HR, ROLES.SUPERUSER],

    // TZ bo'yicha yangi permissions
    can_view_own_data: [
      ROLES.EMPLOYEE, ROLES.MANAGER,
      ROLES.ACCOUNTANT, ROLES.CHIEF_ACCOUNTANT,
      ROLES.SUPERUSER
    ],
    can_edit_department_salary: [ROLES.MANAGER],
    can_submit_to_accountant: [ROLES.MANAGER],
    can_approve_salary: [
      ROLES.ACCOUNTANT,
      ROLES.CHIEF_ACCOUNTANT,
      ROLES.SUPERUSER
    ],
    can_return_to_manager: [
      ROLES.ACCOUNTANT,
      ROLES.CHIEF_ACCOUNTANT,
      ROLES.SUPERUSER
    ],
    can_final_approve: [ROLES.CHIEF_ACCOUNTANT, ROLES.SUPERUSER],
    can_export_to_1c: [ROLES.CHIEF_ACCOUNTANT, ROLES.SUPERUSER],
    can_unlock_documents: [ROLES.SUPERUSER],
    can_view_all_departments: [
      ROLES.ACCOUNTANT,
      ROLES.CHIEF_ACCOUNTANT,
      ROLES.SUPERUSER
    ],
    can_edit_with_comment: [
      ROLES.MANAGER,
      ROLES.ACCOUNTANT,
      ROLES.CHIEF_ACCOUNTANT,
      ROLES.SUPERUSER
    ],
  }

  const allowedRoles = permissionMap[permission] || []
  return allowedRoles.includes(user.role)
}

/**
 * Role-based utility funksiyalar
 */
export const isEmployee = (user) => user?.role === ROLES.EMPLOYEE
export const isManager = (user) => user?.role === ROLES.MANAGER
export const isAccountant = (user) => user?.role === ROLES.ACCOUNTANT
export const isChiefAccountant = (user) => user?.role === ROLES.CHIEF_ACCOUNTANT
export const isSuperuser = (user) => user?.role === ROLES.SUPERUSER || user?.role === ROLES.ADMIN

/**
 * Hodimni tahrirlash huquqi bormi
 *
 * @param {Object} user - Foydalanuvchi
 * @param {Object} employee - Hodim (optional)
 * @returns {boolean}
 */
export const canEditEmployee = (user, employee = null) => {
  if (!user) return false

  // Superuser va Chief Accountant - hammani
  if (isSuperuser(user) || isChiefAccountant(user)) {
    return true
  }

  // Accountant - hammani (izoh bilan)
  if (isAccountant(user)) {
    return true
  }

  // Manager - faqat o'z bo'limidagi hodimlar
  if (isManager(user) && employee) {
    // Employee.department_id === user.managed_department?.id
    return employee.department?.id === user.employee?.department?.id
  }

  // Employee - hech kimni tahrirlolmaydi
  return false
}

/**
 * Hodimni ko'rish huquqi bormi
 *
 * @param {Object} user - Foydalanuvchi
 * @param {Object} employee - Hodim (optional)
 * @returns {boolean}
 */
export const canViewEmployee = (user, employee = null) => {
  if (!user) return false

  // O'zi bo'lsa
  if (employee && user.employee?.id === employee.id) {
    return true
  }

  // Superuser, Chief, Accountant - hammani
  if (isSuperuser(user) || isChiefAccountant(user) || isAccountant(user)) {
    return true
  }

  // Manager - o'z bo'limidagi hodimlar
  if (isManager(user) && employee) {
    return employee.department?.id === user.employee?.department?.id
  }

  // Employee - faqat o'ziniki
  if (isEmployee(user)) {
    return employee && user.employee?.id === employee.id
  }

  return false
}

/**
 * Ish haqini tahrirlash huquqi
 *
 * TZ talabi:
 * - Employee: Yo'q
 * - Manager: Faqat o'z bo'limidagi hodimlar
 * - Accountant: Hamma (izoh bilan)
 * - Chief Accountant: Hamma
 * - Superuser: Hamma
 */
export const canEditSalary = (user, employee = null) => {
  if (!user) return false

  // Employee - yo'q
  if (isEmployee(user)) return false

  // Superuser va Chief - hamma
  if (isSuperuser(user) || isChiefAccountant(user)) return true

  // Accountant - hamma
  if (isAccountant(user)) return true

  // Manager - faqat o'z bo'limi
  if (isManager(user) && employee) {
    return employee.department?.id === user.managed_department?.id
  }

  return false
}

/**
 * Ish haqini ko'rish huquqi
 */
export const canViewSalary = (user, employee = null) => {
  if (!user) return false

  // O'zi bo'lsa
  if (employee && user.employee?.id === employee.id) return true

  // Superuser, Chief, Accountant - hamma
  if (isSuperuser(user) || isChiefAccountant(user) || isAccountant(user)) {
    return true
  }

  // Manager - o'z bo'limi
  if (isManager(user) && employee) {
    return employee.department?.id === user.managed_department?.id
  }

  // Employee - faqat o'ziniki
  if (isEmployee(user)) {
    return employee && user.employee?.id === employee.id
  }

  return false
}

/**
 * Ish haqini tasdiqlash huquqi
 */
export const canApproveSalary = (user) => {
  return hasPermission(user, 'can_approve_salary')
}

/**
 * Yakuniy tasdiqlash va 1C ga yuklash huquqi
 */
export const canFinalApprove = (user) => {
  return hasPermission(user, 'can_final_approve')
}

/**
 * Managerga qaytarish huquqi
 */
export const canReturnToManager = (user) => {
  return hasPermission(user, 'can_return_to_manager')
}

/**
 * 1C ga export qilish huquqi
 */
export const canExportTo1C = (user) => {
  return hasPermission(user, 'can_export_to_1c')
}

/**
 * Barcha bo'limlarni ko'rish huquqi
 */
export const canViewAllDepartments = (user) => {
  return hasPermission(user, 'can_view_all_departments')
}

/**
 * Rolni display qilish uchun label olish
 */
export const getRoleLabel = (role) => {
  return ROLE_LABELS[role] || role
}

/**
 * Rolni display qilish uchun rang olish
 */
export const getRoleColor = (role) => {
  return ROLE_COLORS[role] || 'gray'
}

/**
 * User ma'lumotlarini formatlash
 */
export const formatUser = (user) => {
  if (!user) return null

  return {
    ...user,
    roleLabel: getRoleLabel(user.role),
    roleColor: getRoleColor(user.role),
    isEmployee: isEmployee(user),
    isManager: isManager(user),
    isAccountant: isAccountant(user),
    isChiefAccountant: isChiefAccountant(user),
    isSuperuser: isSuperuser(user),
  }
}

export default {
  ROLES,
  ROLE_LABELS,
  ROLE_COLORS,
  hasRole,
  hasPermission,
  isEmployee,
  isManager,
  isAccountant,
  isChiefAccountant,
  isSuperuser,
  canEditEmployee,
  canViewEmployee,
  canEditSalary,
  canViewSalary,
  canApproveSalary,
  canFinalApprove,
  canReturnToManager,
  canExportTo1C,
  canViewAllDepartments,
  getRoleLabel,
  getRoleColor,
  formatUser,
}

File: backend/apps/core/permissions.py
"""
API Permissions va Decorators.

TZ talabiga mos permission system.
"""

from rest_framework import permissions
from rest_framework.exceptions import PermissionDenied
from functools import wraps
from django.http import JsonResponse


class IsEmployee(permissions.BasePermission):
    """Oddiy xodim huquqi."""

    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.is_employee


class IsManager(permissions.BasePermission):
    """Bo'lim boshlig'i huquqi."""

    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.is_manager


class IsAccountant(permissions.BasePermission):
    """Buxgalter huquqi."""

    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.is_accountant


class IsChiefAccountant(permissions.BasePermission):
    """Glavniy buxgalter huquqi."""

    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.is_chief_accountant


class IsSuperuser(permissions.BasePermission):
    """Administrator huquqi."""

    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.is_superuser_role


class CanEditSalary(permissions.BasePermission):
    """
    Ish haqini tahrirlash huquqi.

    TZ talabi:
    - Employee: Yo'q
    - Manager: Faqat o'z bo'limidagi hodimlar
    - Accountant: Hamma (izoh bilan)
    - Chief Accountant: Hamma
    - Superuser: Hamma
    """

    def has_object_permission(self, request, view, obj):
        """
        Obyekt darajasida permission tekshirish.

        obj - bu SalaryCalculation yoki Employee obyekti bo'lishi mumkin.
        """
        user = request.user

        # GET request - ko'rish huquqi
        if request.method == 'GET':
            if hasattr(obj, 'employee'):
                employee = obj.employee
            else:
                employee = obj

            return user.can_view_salary_for_employee(employee)

        # PUT/PATCH/DELETE - tahrirlash huquqi
        if request.method in ['PUT', 'PATCH', 'DELETE']:
            if hasattr(obj, 'employee'):
                employee = obj.employee
            else:
                employee = obj

            return user.can_edit_salary_for_employee(employee)

        return False


class CanApproveSalary(permissions.BasePermission):
    """
    Ish haqini tasdiqlash huquqi.

    Faqat Accountant, Chief Accountant va Superuser.
    """

    def has_permission(self, request, view):
        user = request.user
        return user.is_authenticated and (
            user.is_accountant or
            user.is_chief_accountant or
            user.is_superuser_role
        )


class CanFinalApprove(permissions.BasePermission):
    """
    Yakuniy tasdiqlash va 1C ga yuklash huquqi.

    Faqat Chief Accountant va Superuser.
    """

    def has_permission(self, request, view):
        user = request.user
        return user.is_authenticated and (
            user.is_chief_accountant or
            user.is_superuser_role
        )


# Decorator funksiyalar

def require_role(*roles):
    """
    View uchun rol talab qiladigan decorator.

    Usage:
        @require_role('manager', 'accountant')
        @api_view(['GET'])
        def my_view(request):
            ...

    Args:
        *roles: Ruxsat berilgan rollar ro'yxati
    """
    def decorator(view_func):
        @wraps(view_func)
        def wrapper(request, *args, **kwargs):
            if not request.user.is_authenticated:
                return JsonResponse({
                    'error': 'Authentication required'
                }, status=401)

            user_role = request.user.role
            if user_role not in roles and not request.user.is_superuser_role:
                return JsonResponse({
                    'error': f'Permission denied. Required roles: {", ".join(roles)}'
                }, status=403)

            return view_func(request, *args, **kwargs)
        return wrapper
    return decorator


def require_permission(permission_name):
    """
    View uchun permission talab qiladigan decorator.

    Usage:
        @require_permission('can_edit_salary')
        @api_view(['POST'])
        def edit_salary(request):
            ...

    Args:
        permission_name: Permission nomi
    """
    def decorator(view_func):
        @wraps(view_func)
        def wrapper(request, *args, **kwargs):
            if not request.user.is_authenticated:
                return JsonResponse({
                    'error': 'Authentication required'
                }, status=401)

            if not request.user.has_permission(permission_name):
                return JsonResponse({
                    'error': f'Permission denied: {permission_name}'
                }, status=403)

            return view_func(request, *args, **kwargs)
        return wrapper
    return decorator


def check_department_access(view_func):
    """
    Manager uchun faqat o'z bo'limiga dostup beruvchi decorator.

    Usage:
        @check_department_access
        @api_view(['GET'])
        def department_employees(request, department_id):
            # department_id o'z bo'limiga tegishli ekanligini tekshiradi
            ...
    """
    @wraps(view_func)
    def wrapper(request, *args, **kwargs):
        if not request.user.is_authenticated:
            return JsonResponse({
                'error': 'Authentication required'
            }, status=401)

        # Superuser va Chief/Accountant lar hammaga dostup
        if (request.user.is_superuser_role or
            request.user.is_chief_accountant or
            request.user.is_accountant):
            return view_func(request, *args, **kwargs)

        # Manager - faqat o'z bo'limi
        if request.user.is_manager:
            department_id = kwargs.get('department_id') or kwargs.get('pk')
            if department_id:
                managed_dept = request.user.managed_department
                if managed_dept and str(managed_dept.id) == str(department_id):
                    return view_func(request, *args, **kwargs)
                else:
                    return JsonResponse({
                        'error': 'You can only access your own department'
                    }, status=403)

        return JsonResponse({
            'error': 'Permission denied'
        }, status=403)

    return wrapper


def require_comment_for_edit(view_func):
    """
    Tahrirlash uchun comment majburiy qiladigan decorator.

    TZ talabi: Buxgalter va Managerlar tahrirlashda izoh qoldirishi shart.

    Usage:
        @require_comment_for_edit
        @api_view(['PUT', 'PATCH'])
        def update_salary(request, pk):
            comment = request.data.get('comment')
            # comment mavjudligini tekshiradi
            ...
    """
    @wraps(view_func)
    def wrapper(request, *args, **kwargs):
        # Faqat PUT/PATCH metodlari uchun
        if request.method in ['PUT', 'PATCH']:
            comment = request.data.get('comment') or request.data.get('notes')

            # Superuser uchun comment majburiy emas (lekin tavsiya etiladi)
            if request.user.is_superuser_role:
                return view_func(request, *args, **kwargs)

            # Boshqa rollar uchun comment MAJBURIY
            if not comment or not comment.strip():
                return JsonResponse({
                    'error': 'Comment is required for editing (TZ requirement)',
                    'field': 'comment'
                }, status=400)

        return view_func(request, *args, **kwargs)

    return wrapper


class RoleBasedPermission(permissions.BasePermission):
    """
    Universal role-based permission class.

    Usage in ViewSet:
        permission_classes = [RoleBasedPermission]
        role_permissions = {
            'list': ['employee', 'manager', 'accountant', 'chief_accountant', 'superuser'],
            'retrieve': ['employee', 'manager', 'accountant', 'chief_accountant', 'superuser'],
            'create': ['manager', 'accountant', 'chief_accountant', 'superuser'],
            'update': ['manager', 'accountant', 'chief_accountant', 'superuser'],
            'destroy': ['superuser'],
        }
    """

    def has_permission(self, request, view):
        """Permission tekshirish."""
        if not request.user or not request.user.is_authenticated:
            return False

        # ViewSet'da role_permissions mavjudligini tekshirish
        if not hasattr(view, 'role_permissions'):
            # Agar belgilanmagan bo'lsa, authenticated user lar uchun ruxsat
            return True

        # Action nomini olish
        action = view.action if hasattr(view, 'action') else None
        if not action:
            return True

        # Action uchun ruxsat berilgan rollarni olish
        allowed_roles = view.role_permissions.get(action, [])

        # Superuser har doim ruxsat oladi
        if request.user.is_superuser_role:
            return True

        # Foydalanuvchi roli ruxsat berilgan rollar ichida bormi
        return request.user.role in allowed_roles


def log_action(action_name):
    """
    API action'ni audit log'ga yozadigan decorator.

    Usage:
        @log_action('salary_updated')
        @api_view(['PUT'])
        def update_salary(request, pk):
            ...
    """
    def decorator(view_func):
        @wraps(view_func)
        def wrapper(request, *args, **kwargs):
            # View'ni bajarish
            response = view_func(request, *args, **kwargs)

            # Audit log yozish (agar success bo'lsa)
            if response.status_code in [200, 201, 204]:
                try:
                    from apps.core.models import log_model_change

                    # Request'dan ma'lumotlarni olish
                    ip_address = request.META.get('REMOTE_ADDR')
                    user_agent = request.META.get('HTTP_USER_AGENT')
                    comment = request.data.get('comment', f'Action: {action_name}')

                    # Log yozish (instance kerak bo'ladi, shuning uchun bu yerda faqat marker)
                    # Actual logging view ichida amalga oshiriladi
                    request._audit_action = action_name
                    request._audit_ip = ip_address
                    request._audit_user_agent = user_agent

                except Exception as e:
                    # Audit log xatosi asosiy jarayonga ta'sir qilmasligi kerak
                    pass

            return response
        return wrapper
    return decorator
